wrapper/src/main/java/software/amazon/jdbc/authentication/AwsCredentialsService.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc.authentication;
+
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
+import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
+
+
+public class AwsCredentialsService {
+  static Supplier<AwsCredentialsProvider> customSupplier = null;
+  static int cacheTimeout = 0;
+  static TimeUnit timeoutUnit;
+  static long lastRefreshTimeNano;
+  static AwsCredentialsProvider PROVIDER_CACHE = null;
+
+  public static void setCustomSupplier(Supplier<AwsCredentialsProvider> customSupplier) {
+    AwsCredentialsService.customSupplier = customSupplier;
+    clearCache();
+  }
+
+  public static void setCustomSupplier(Supplier<AwsCredentialsProvider> customSupplier, int timeout,
+      TimeUnit timeoutUnits) {
+    AwsCredentialsService.customSupplier = customSupplier;
+    AwsCredentialsService.cacheTimeout = timeout;
+    AwsCredentialsService.timeoutUnit = timeoutUnits;
+    PROVIDER_CACHE = null;
+  }
+
+  public static void resetCustomSupplier() {
+    customSupplier = null;
+    clearCache();
+  }
+
+  private static void clearCache() {
+    cacheTimeout = 0;
+    timeoutUnit = null;
+    PROVIDER_CACHE = null;
+  }
+
+  public static AwsCredentialsProvider getProvider() {
+    if (PROVIDER_CACHE == null || cacheTimeout == 0 || timeoutUnit == null) {
+      return refreshProvider();
+    }
+
+    long timeSinceLastRefreshNano = System.nanoTime() - lastRefreshTimeNano;
+    if (timeSinceLastRefreshNano >= timeoutUnit.toNanos(cacheTimeout)) {
+      return refreshProvider();
+    } else {
+      return PROVIDER_CACHE;
+    }
+  }
+
+  private static AwsCredentialsProvider refreshProvider() {
+    AwsCredentialsProvider provider;
+    if (customSupplier != null) {
+      provider = customSupplier.get();
+    } else {
+      provider = getDefaultProvider();
+    }
+
+    lastRefreshTimeNano = System.nanoTime();
```

Comment:
If cache isn't enabled then it shouldn't store the provider in PROVIDER_CACHE.

Reply:
If cache is enabled, then getting a new provider and storing it to PROVIDER_CACHE should be atomic and synchronized.

Reply:
https://www.section.io/engineering-education/implementing-the-singleton-pattern-in-c-sharp/#:~:text=3.%20Using%20double%2Dcheck%20locking%20to%20ensure%20thread%20safety%20in%20a%20singleton